lib/rucio/core/authentication.py

```diff
@@ -384,17 +1331,82 @@ def query_token(token, session=None):
     if ret:
         return {'account': ret[0][0],
                 'identity': ret[0][1],
-                'lifetime': ret[0][2]}
+                'lifetime': ret[0][2],
+                'audience': ret[0][3],
+                'authz_scope': ret[0][4]}
     return None
+
+
+@read_session
+def validate_jwt(json_web_token, session=None):
+    """
+    Verifies signature and validity of a JSON Web Token.
+    Gets the issuer public keys from the OIDC_client
+    and verifies the validity of the token.
+
+    :param json_web_token: the JWT string to verify
+
+    :returns: dictionary { account: <account name>,
+                           identity: <identity>,
+                           lifetime: <token lifetime>,
+                           audience: <audience>,
+                           authz_scope: <authz_scope> }
+              if successful, None otherwise.
+    """
+    global OIDC_clients
+    try:
+        # getting issuer from the token payload
+        issuer = JWT().unpack(json_web_token).payload()['iss']
+        OIDC_client = OIDC_clients[issuer]
+        issuer_keys = OIDC_client.keyjar.get_issuer_keys(issuer)
+        jsig = JWS()
+        claim_dict = jsig.verify_compact(json_web_token, issuer_keys)
+        identity_string = oidc_identity_string(claim_dict['sub'], claim_dict['iss'])
+        expiry_date = datetime.datetime.utcfromtimestamp(claim_dict['exp'])
+        if expiry_date < datetime.datetime.utcnow():  # check if expired
+            return None
+        scope = None
+        if 'scope' in claim_dict:
+            scope = claim_dict['scope']
+        audience = claim_dict.get('aud', None)
+        # if there is no audience and scope information, try to get it from IdP introspection endpoint
+        if audience is None or scope is None:
+            clprocess = subprocess.Popen(['curl', '-s', '-L', '-u', '%s:%s' % (OIDC_client.client_id, OIDC_client.client_secret), '-d', 'token=%s' % (json_web_token), OIDC_client.introspection_endpoint], shell=False, stdout=subprocess.PIPE)
+            inspect_claims = json.loads(clprocess.communicate()[0])
+            try:
+                audience = inspect_claims['aud']
+                scope = inspect_claims['scope']
+            except:
+                pass
+        # this assumes token has been previously looked up in DB before to be sure that we do not have the right account already in the DB !
+        account = get_default_account(identity_string, IdentityType.OIDC, True, session=session)
+        # account = session.query(models.Token.account).filter(models.Token.identity == identity_string).first()
+        value = {'account': account,
+                 'identity': identity_string,
+                 'lifetime': expiry_date,
+                 'audience': audience,
+                 'authz_scope': scope}
+        record_counter(counters='JSONWebToken.valid')
+        return value
+    except:
+        record_counter(counters='JSONWebToken.invalid')
+        return None
```

Comment:
very large file now, move to `core/oidc.py`

Reply:
I made it 3 files:  `authentication.py, oidc.py `and `tokens.py`. And I moved the `query_bunches` and `filter_thread_load` functions into common.utills. In the `tokens.py` there is [1], i.e. all that has something to do with general token validation and manipulation.
 I changed the imports in other files accordingly ...

What do you think ? 

[1] token_key_generator, TOKENREGION, save_validated_token, delete_expired_tokens, get_tokens_for_refresh, change_refresh_state, query_token, validate_auth_token, validate_jwt

Reply:
I like the separation; Very nice 👍